rfcs/search_relevancy_sandbox/20230331-project_proposal_search_relevancy_sandbox.md

```diff
@@ -0,0 +1,145 @@
+test API results from all currently available providers.
+
+### Production data volume
+
+The production-data-volume index will be an index of similar size to production.
+This is how the data refresh is currently set up and does not require any
+additional work to set up.
+
+This index will allow us to test how queries and indexing operations perform at
+a data volume consistent with production.
+
+## Success
+
+<!-- How do we measure the success of the project? How do we know our ideas worked? -->
+
+This project can be considered a success once any maintainer can make
+adjustments to the staging API database & Elasticsearch index based on the
+requirements described above.
+
+## Participants and stakeholders
+
+<!-- Who is working on the project and who are the external stakeholders, if any? Consider the lead, implementers, designers, and other stakeholders who have a say in how the project goes. -->
+
+- Lead: @AetherUnbound
+- Implementation:
+  - @AetherUnbound
+  - TBD
+- Stakeholders:
+  - Openverse Team
+
+## Infrastructure
+
+<!-- What infrastructural considerations need to be made for this project? If there are none, say so explicitly rather than deleting the section. -->
+
+This project may require infrastructure work necessary to assist with the rapid
+iteration of the data refresh server. Further details will be determined in the
+implementation plan.
+
+The project will also likely involve the creation or modification of Airflow
+DAGs related to the data refresh and infrastructure modification. This will
+require interfacing Airflow with the staging data refresh server, API database,
+and Elasticsearch cluster.
+
+## Accessibility
+
+<!-- Are there specific accessibility concerns relevant to this project? Do you expect new UI elements that would need particular care to ensure they're implemented in an accessible way? Consider also low-spec device and slow internet accessibility, if relevant. -->
+
+Not every member of the maintainer team is intimately familiar with Airflow - we
+will need to provide clear instructions for maintainers on how to use Airflow to
+run any of the above mechanisms.
+
+## Marketing
+
+<!-- Are there potential marketing opportunities that we'd need to coordinate with the community to accomplish? If there are none, say so explicitly rather than deleting the section. -->
+
+No marketing is required as this is an improvement internal to the team.
+
+## Required implementation plans
+
+<!-- What are the required implementation plans? Consider if they should be split per level of the stack or per feature. -->
+
+In the order they should be completed:
+
+1. Staging API update procedure
+   - This plan will describe a DAG which can be triggered to update the staging
+     API database.
+2. Rapid iteration on data refresh server index configuration
+   - This plan will describe how rapid iteration of the Elasticsearch index
+     configuration can happen.
```

Comment:
What does "the Elasticsearch index" mean? Specifically the "the" isn't clear to me :thinking: ES can have any number of indexes, each with their own independent configuration. Rapidly iterating on index configuration could be as easy as changing the request body to the create index endpoint, which someone could either issue directly in Elasticvue (or similar) or give as input to an Airflow DAG, for example.

I also think we should leverage the re-index API for this. I could see an Airflow DAG meant for iterating on index configuration that works something like this:

1. Trigger the DAG with a new index configuration body.
    - The existing configuration can be generated using the `index_settings` function from the ingestion server project. This function runs fine in a Python shell with zero setup inside the ingestion_server project.
1. The DAG creates the index via the create index endpoint, either using a name provided when the DAG was triggered or by generating a UUID or date-based suffix. This should happen by hitting ES directly, not through the ingestion server.
1. The DAG then calls the reindex API to copy the data from the existing (periodically refreshed?) "production data" index into the new index. If we want to incorporate "proportional data" or have a subset of data, we could also have the DAG take in a query. If we just need to test a stemmer change, for example, we can query for documents that include the terms in question. For broader text analysis changes we'd want to use the full production data (see my comment above). If we do use the proportional-by-provider data, we can allow changing the origin index between the full production data and the proportional-by-provider index.
1. After reindex is done, refresh the new index.

Done! We've just completed one index configuration iteration.

This bypasses the complexity of the ingestion server and wouldn't require any changes to the ingestion server _at all_.

Reply:
I love the idea of leveraging Airflow the way you're describing!

By "the" index I mean - the primary index that will eventually serve as the index behind the `image|audio` alias. In this case/example, that's not what would be iterated on though, I'll try to reword this to be more generic as it relates to indices. The specifics you describe will go into the eventual implementation plan for this step, so I won't include them here.

Reply:
Sounds good. I think it is worth recording the general approach in this document, otherwise the person doing the implementation plan may not know about this being a viable approach and go down a different rabbit hole. A small note in the implementation plan section that we can use the reindex API directly from Airflow to ES rather than passing through the ingestion server would be sufficient. Right now the suggested approach feels heavily skewed towards needing to iterate on the ingestion server code itself and maybe even moving it to ECS, which is a _lot_ more work that we could avoid. I think it would at least triple the scope of the implementation plan if it included moving the ingestion server to ECS vs just hitting ES directly from Airflow and using the reindex API.